docs/proposals/api-types-package-structure.md

```diff
@@ -0,0 +1,475 @@
+do RPCs to look up other cluster state is allowed.
+
+### Package: `apis/group/v1/defaulting`
+
+This package contains defaulting functions, and exports a Register() function
+which will register them (but does not register them as a side effect of being
+imported).
+
+The functions apply to only the `group/v1/types.go` types, *not* the parent
+`group/types.go` types.
+
+High-level requirements:
+* Any tags that autogeneration pays attention to (e.g., `// +default: xxxx`)
+  should be readable by humans as well as machines.
+* Things that require a default should not compile until the defaulting function
+  is supplied.
+* Defaults may need to be hirearchical-- e.g., the container resources in a pod
+  temaplate are defaulted differently than the resources in a pod. The simplest
+  way to do this is to set the defaults in the function for the template and
+  separately for the pod, instead of setting them on the container's resources.
+
+The allowed imports are constrained; in particular, nothing that would let you
+do RPCs to look up other cluster state is allowed. Defaulting that requires
+information not contained in the object is not allowed; you must change your
+design until it's not needed.
+
+### Package: `apis/group/v1/validation`
+
+This package contains validation functions, and if imported, will register them.
+
+The functions apply to only the `group/v1/types.go` types, *not* the parent
+`group/types.go` types.
+
+The allowed imports are constrained; in particular, nothing that would let you
+do RPCs to look up other cluster state is allowed. Validation functions validate
+*only* the object, not cluster-wide or other constraints, which must be validate
+elsewhere.
+
+There are two sorts of validation functions, those that can be automatically
+written and those that need human attention.
+
+The former will be autogenerated by looking for directives in the comments
+preceding the types or their embedded struct members. Allowed directives will
+include:
+* `// +validation=minmax{<numeric value>, <numeric value>}`
+* `// +validation=regexp/<regexp>/`
+* `// +validation=<exact value>`
+
+If no validation directive is in the comments, a stub with a
+`panic("write me!")` marker will be generated. To prevent this, a
+`// +validation=none` directive may be added to the comment.
+
+(We can hash out the exact form of the tags later; it should be easy for human
+readers to figure out what will happen.)
+
+### Package: `apis/group/v1beta1`
+
+This package layout is the same as `apis/group/v1`, but it is an older version
+of the API group, which may lack features.
+
+### Package: `apis/group/v2alpha1`
+
+This package layout is the same as `apis/group/v1`, but it is an newer version
+of the API group, which may have extra features. It may have types for an
+additional resource, for example.
+
+It is a rule that a client should never have to pay attention to multiple
+versions of a group at the same time, so v2alpha1 copies all types from v1, even
```

Comment:
How this applies to your suggestion captured in https://github.com/kubernetes/kubernetes/issues/37166?